services/api/service.go

```diff
@@ -835,6 +928,71 @@ func (api *RelayAPI) handleGetPayload(w http.ResponseWriter, req *http.Request)
 		if err != nil {
 			log.WithError(err).Error("failed to increment builder-stats after getPayload")
 		}
+
+		builderStatus, err := api.redis.GetBlockBuilderStatus(bidTrace.BuilderPubkey.String())
+		if err != nil {
+			log.WithError(err).Error("failed to get builder status")
+		}
+
+		// Check if the block was valid if the builder is OptimisticActive or OptimisticDemoted.
+		if builderStatus == common.OptimisticActive || builderStatus == common.OptimisticDemoted {
+			// Set to locked while we process the winning block.
+			api.setStatusByCollateralID(bidTrace.BuilderPubkey.String(), common.OptimisticLocked)
+			submitBlockReq := types.BuilderSubmitBlockRequest{
+				Signature:        payload.Signature,
+				Message:          &bidTrace.BidTrace,
+				ExecutionPayload: getPayloadResp.Data,
+			}
+
+			simErr := api.simulateBlock(blockSimOptions{
+				ctx:      req.Context(),
+				log:      log,
+				highPrio: true, // manually set to true for these blocks.
+				req: &BuilderBlockValidationRequest{
+					BuilderSubmitBlockRequest: submitBlockReq,
+				},
+			})
+			if simErr != nil {
+				signedBeaconBlock := SignedBlindedBeaconBlockToBeaconBlock(payload, getPayloadResp.Data)
+				log = log.WithFields(logrus.Fields{
+					"bidTrace":          bidTrace,
+					"signedBeaconBlock": signedBeaconBlock,
```

Comment:
ya i think i was concerned about the case that Chris described, which is if the insert into the DB fails. in that case we really want to be sure we don't lose the data about the invalid block. i moved that addition to the logger down to the error handling of the UpsertBuilderDemotion. that way it will only ever happen if the DB write fails. does that seem reasonable?

Reply:
Right, it makes a lot of sense to use logs as a makeshift fallback DB.